AnyText/AnyText.Lsp/LspServer.SemanticToken.cs

```diff
@@ -0,0 +1,221 @@
+using LspTypes;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NMF.AnyText
+{
+    public partial class LspServer
+    {
+        private readonly Dictionary<string, uint[]> _previousSemanticTokens = new();
+
+        /// <summary>
+        ///     Handles the <c>textDocument/semanticTokens/full</c> request from the client. This is used to retrieve all semantic
+        ///     tokens for a document.
+        /// </summary>
+        /// <param name="arg">The JSON token containing the parameters of the request. (SemanticTokensParams)</param>
+        /// <returns>A <see cref="SemanticTokens" /> object containing the full set of semantic tokens for the document.</returns>
+        public SemanticTokens QuerySemanticTokens(JToken arg)
+        {
+            var semanticTokensParams = arg.ToObject<SemanticTokensParams>();
+            var uri = semanticTokensParams.TextDocument.Uri;
+            
+            if (!_documents.TryGetValue(uri, out var document))
+                return new SemanticTokens { ResultId = null, Data = Array.Empty<uint>() };
+
+            var semanticElements = document.GetSemanticElementsFromRoot().ToArray();
+
+            var resultId = Guid.NewGuid().ToString();
+            _previousSemanticTokens[resultId] = semanticElements;
+
+            return new SemanticTokens
+            {
+                ResultId = resultId,
+                Data = semanticElements
+            };
+        }
+
+        /// <summary>
+        ///     Handles the <c>textDocument/semanticTokens/full/delta</c> request from the client. This is used to retrieve only
+        ///     the changes (delta) in semantic tokens for a document.
+        /// </summary>
+        /// <param name="arg">The JSON token containing the parameters of the request. (SemanticTokensDeltaParams)</param>
+        /// <returns>A <see cref="SemanticTokensDelta" /> object containing only the delta of semantic tokens for the document.</returns>
+        public SemanticTokensDelta QuerySemanticTokensDelta(JToken arg)
+        {
+            var semanticTokensParams = arg.ToObject<SemanticTokensDeltaParams>();
+            var uri = semanticTokensParams.TextDocument.Uri;
+
+            if (!_documents.TryGetValue(uri, out var document))
+                return new SemanticTokensDelta { ResultId = null, Edits = Array.Empty<SemanticTokensEdit>() };
+            _previousSemanticTokens.TryGetValue(semanticTokensParams.PreviousResultId, out var previousTokens);
+            
+            var currentTokens = document
+                .GetSemanticElementsFromRoot().ToArray();
+            _previousSemanticTokens[semanticTokensParams.PreviousResultId] = currentTokens;
+
+            var originalLength = previousTokens?.Length ?? 0;
+            var modifiedLength = currentTokens.Length;
+
+            var startIndex = 0;
+
+            // Find the matching prefix
```

Comment:
Ich glaube jetzt wäre noch ein guter Zeitpunkt für Refactorings, also alle Bereiche, die sowieso schon einen extra überschriebenen Kommentar haben, einfach in private Hilfsmethoden zu kapseln, die genau so heißen.